### ISSUE TITLE
Discrepencies in cell_count between flagser and flagser-count

### ISSUE THREAD

**Issue by gtauzin** [luetge/flagser#27]
Hi,

When binding flagser-count, and developing tests in [pyflagser](https://github.com/giotto-ai/pyflagser), I have noticed an inconsistency between the cell count results obtained by running `flagser` and the one obtained by running `flagser-count` for two of the example files: `d10.flag` and `medium-test-data.flag`. For all other example files in `test/`, both executable return the same cell count.

Taking the example of `d10.flag`, we have:

#### `flagser` cell counts (dim 0 to 10)
10 90 720 5040 30240 151200 604800 1814400 3628800 3628800 0

#### `flagser-count` cell counts (dim 0 to 9)
8 72 576 4032 24192 120960 483840 1451520 2903040 2903040

Just looking at the number of vertices, it seems that `flagser-count` is the one outputting wrong results. The euler characteristic is also off as a consequence.

**Comment by luetge** [comment]
Hm, that's strange, when I run it on my machine (latest master) I get this:

```bash
daniel ~/programming/flagser master
$ flagser-count test/d10.flag
# [euler_characteristic cell_count_dim_0 cell_count_dim_1 ...]
-1334960 10 90 720 5040 30240 151200 604800 1814400 3628800 3628800
daniel ~/programming/flagser master
$ flagser test/d10.flag --out /tmp/abc.homology && cat /tmp/abc.homology && rm /tmp/abc.homology
The dimensions of the mod-2 homology of the full complex are:

dim H_0 = 1
dim H_1 = 0
dim H_2 = 0
dim H_3 = 0
dim H_4 = 0
dim H_5 = 0
dim H_6 = 0
dim H_7 = 0
dim H_8 = 0
dim H_9 = 1334961
The Euler characteristic is given by: -1334960
# Cell counts (of dimensions between 0 and 10):
10 90 720 5040 30240 151200 604800 1814400 3628800 3628800 0
# Euler characteristic:
-1334960
# Betti numbers (between 0 and 9):
1 0 0 0 0 0 0 0 0 1334961
```

Can you maybe provide some more details on how to reproduce it? (System, compiler, ...)

Thanks!

**Comment by gtauzin** [comment]
Oh indeed, that's strange. That's what I get (also latest master)  running `flagser-count` compiled with gcc-6.4.0 with (or without HDF5) on a Ubuntu 14.04 LTS laptop. 

```shell
~/workspace/luetge/flagser$ ./flagser-count test/d10.flag
# [euler_characteristic cell_count_dim_0 cell_count_dim_1 ...]
-1067968 8 72 576 4032 24192 120960 483840 1451520 2903040 2903040
```

Moreover, the CI of `pyflagser` builds `flagser` and compares the cell_count of `flagser` to the cell_count of `flagser-count` on MacOS 2014, Windows 2016, and other linux distros.  All pipeline fail for the same reason and the log can be found [here](https://dev.azure.com/maintainers/Giotto/_build/results?buildId=2113&view=logs&j=bd2b5cb2-63a7-5df6-c258-4f0eed6d0bea&t=932fc33d-35ad-54a3-ce50-95180cc53946). 

Additionally, I tried on a cluster running on CentOS Linux release 7.1.1503 and got the same wrong result too.

**Comment by MonkeyBreaker** [comment]
Hi !

Let me add my observations to the discussions.
I can reproduce the same observations as @gtauzin, I have inconsistencies between `flagser` and `flagser-count`.

Looking why this is possible, I just found out that the number of `cell_count` depends on the value of `PARALLEL_THREADS`. If On my machine I change `PARALLEL_THREADS` from 8 to 10, I got the following results:

```bash
# [euler_characteristic cell_count_dim_0 cell_count_dim_1 ...]
-1334960 10 90 720 5040 30240 151200 604800 1814400 3628800 3628800
```

Which seems to be the correct results. 
Changing `PARALLEL_THREADS` does not change the results for `flagser`, they keep consistent.

If I have some time I'll try to go deeper on why, but at the moment I don't have more time to spend on the issue, sorry for that.

Best,
Julián

**Comment by gtauzin** [comment]
Well spotted, thanks @MonkeyBreaker !

I can reproduce the behavior you described and it seemed that whenever the number of vertices is higher than ``PARALLEL_THREADS`` it just compute a partial count stemming from ``PARALLEL_THREADS`` number of vertices.

**Comment by luetge** [comment]
Great, that should be easy to fix, will have a look. Thanks a lot!

**Comment by luetge** [comment]
Fixed in #28, really confusing that this code works on my machine at all.. Thanks for pointing it out!
